src/core/solvers/ipopt/multiple-shooting-nlp.cpp

```diff
@@ -0,0 +1,589 @@
+
+        state_->integrate(xs_[i], data_.dx, data_.x);
+
+        problem_->get_runningModels()[i]->calc(problem_->get_runningDatas()[i], data_.x, data_.control);
+
+        obj_value += problem_->get_runningDatas()[i]->cost;
+    }
+
+    // Terminal costs
+    data_.dx = Eigen::VectorXd::Map(x + T_ * (ndx_ + nu_), ndx_);
+
+    state_->integrate(xs_[T_], data_.dx, data_.x);
+
+    problem_->get_terminalModel()->calc(problem_->get_terminalData(), data_.x);
+    obj_value += problem_->get_terminalData()->cost;
+
+    return true;
+}
+// [TNLP_eval_f]
+
+// [TNLP_eval_grad_f]
+// return the gradient of the objective function grad_{x} f(x)
+bool MultipleShootingNlp::eval_grad_f(Ipopt::Index n, const Ipopt::Number *x, bool new_x, Ipopt::Number *grad_f)
+{
+    assert(n == nvar_);
+
+    for (size_t i = 0; i < T_; i++) {
+        data_.dx      = Eigen::VectorXd::Map(x + i * (ndx_ + nu_), ndx_);
+        data_.control = Eigen::VectorXd::Map(x + i * (ndx_ + nu_) + ndx_, nu_);
+
+        state_->integrate(xs_[i], data_.dx, data_.x);
+
+        state_->Jintegrate(xs_[i], data_.dx, data_.Jsum_x, data_.Jsum_dx, crocoddyl::second, crocoddyl::setto);
+        problem_->get_runningModels()[i]->calcDiff(problem_->get_runningDatas()[i], data_.x, data_.control);
+        data_.Ldx = data_.Jsum_dx.transpose() * problem_->get_runningDatas()[i]->Lx;
+
+        for (size_t j = 0; j < ndx_; j++) {
+            grad_f[i * (ndx_ + nu_) + j] = data_.Ldx(j);
+        }
+
+        for (size_t j = 0; j < nu_; j++) {
+            grad_f[i * (ndx_ + nu_) + ndx_ + j] = problem_->get_runningDatas()[i]->Lu(j);
+        }
+    }
+
+    data_.dx = Eigen::VectorXd::Map(x + T_ * (ndx_ + nu_), ndx_);
+
+    state_->integrate(xs_[T_], data_.dx, data_.x);
+    state_->Jintegrate(xs_[T_], data_.dx, data_.Jsum_x, data_.Jsum_dx, crocoddyl::second, crocoddyl::setto);
+
+    problem_->get_terminalModel()->calcDiff(problem_->get_terminalData(), data_.x);
+    data_.Ldx = data_.Jsum_dx.transpose() * problem_->get_terminalData()->Lx;
+
+    for (size_t j = 0; j < ndx_; j++) {
+        grad_f[T_ * (ndx_ + nu_) + j] = data_.Ldx(j);
+    }
+
+    return true;
+}
+// [TNLP_eval_grad_f]
+
+// [TNLP_eval_g]
+// return the value of the constraints: g(x)
+bool MultipleShootingNlp::eval_g(Ipopt::Index n, const Ipopt::Number *x, bool new_x, Ipopt::Index m, Ipopt::Number *g)
+{
+    assert(n == nvar_);
+    assert(m == nconst_);
+
+    // Dynamic constraints
+    for (size_t i = 0; i < T_; i++) {
+        data_.dx      = Eigen::VectorXd::Map(x + i * (ndx_ + nu_), ndx_);
+        data_.dxnext  = Eigen::VectorXd::Map(x + (i + 1) * (ndx_ + nu_), ndx_);
+        data_.control = Eigen::VectorXd::Map(x + i * (ndx_ + nu_) + ndx_, nu_);
+
+        state_->integrate(xs_[i], data_.dx, data_.x);
+        state_->integrate(xs_[i + 1], data_.dxnext, data_.xnext);
+
+        problem_->get_runningModels()[i]->calc(problem_->get_runningDatas()[i], data_.x, data_.control);
```

Comment:
Are you sure we need to run `calc`?
I suspect this is not necessary as we run it for the cost computation.